stan/math/prim/fun/eval.hpp

```diff
@@ -17,8 +17,20 @@ namespace math {
  **/
 template <typename T,
           require_same_t<std::decay_t<T>, plain_type_t<T>>* = nullptr>
-inline decltype(auto) eval(T&& arg) {
-  return std::forward<T>(arg);
+inline T eval(T&& arg) {
+  return std::move(arg);
+}
+
+template <typename T,
+          require_same_t<std::decay_t<T>, plain_type_t<T>>* = nullptr>
+inline const T& eval(const T& arg) {
+  return arg;
+}
+
+template <typename T,
+          require_same_t<std::decay_t<T>, plain_type_t<T>>* = nullptr>
+inline T& eval(T& arg) {
+  return arg;
```

Comment:
Why did you feel you need to write these out? You could just replace `decltype(auto)` with `T`.

Reply:
Ah the copy ellision thing was intuitive to me so I wrote it by hand but I think I got this now so I'll fix these up

https://godbolt.org/z/a3TMPW